### PR TITLE
Multi-file support

### PR DESCRIPTION
**Pull request by g-r-a-n-t**
### What was wrong?

There's no multi-file (or module) support.

### How was it fixed?

Mostly copied Rust's crate system, but use the the term *ingot* instead of crate. 

Below is an example of an ingot's file tree, as supported by the current implementation.

```
`-- basic_ingot
    `-- src
        |-- bar
        |   `-- baz.fe
        |-- bing.fe
        |-- ding
        |   |-- dang.fe
        |   `-- dong.fe
        `-- main.fe
```

When compared to the file tree of a crate, there are a few things missing: 

- source files accompanying each directory module (e.g. `my_mod.fe`) 
- support for `lib.fe` files in the `src` dir
- configuration file
- test directory

There are also a few things missing in the language: 
- module-level `pub` modifier, so all items in a module are public
- `mod` statements, so all files in the tree are accessible

The last missing thing to note is that ingots cannot be compiled with each other, so we can't start building prelude libraries quite yet.

---

In general, I think that being able to provide a set of prelude libraries is most desirable, which means that the missing features mentioned above should be implemented in an order that allow us to start doing that ASAP.

### To-Do

[//]: # (Stay ahead of things, add list items here!)
- [ ] OPTIONAL: Update [Spec](https://github.com/ethereum/fe/blob/master/docs/src/spec/index.md) if applicable
- [ ] Add entry to the [release notes](https://github.com/ethereum/fe/blob/master/newsfragments/README.md) (may forgo for trivial changes)
- [x] Clean up commit history

### CODE REVIEW

File: crates/analyzer/tests/analysis.rs [hunk 1]
```diff
@@ -1,341 +1,341 @@
-use fe_analyzer::context;
```
Comment (g-r-a-n-t) [inline crates/analyzer/tests/analysis.rs:1]: get these working again

File: crates/analyzer/tests/errors.rs [hunk 2]
```diff
@@ -1,258 +1,258 @@
-//! Tests for contracts that should cause compile errors
```
Comment (g-r-a-n-t) [inline crates/analyzer/tests/errors.rs:1]: add some error tests for non-existent modules

File: crates/yulc/src/lib.rs [hunk 3]
```diff
@@ -1,38 +1,29 @@
...
-pub type Bytecode = String;
-/// A mapping of contract names and their bytecode.
-pub type NamedBytecodeContracts = HashMap<ContractName, Bytecode>;
+use indexmap::map::IndexMap;
 
 #[derive(Debug)]
 pub struct YulcError(pub String);
 
 /// Compile a map of Yul contracts to a map of bytecode contracts.
+// TODO: make the output format clear
```
Comment (g-r-a-n-t) [inline crates/yulc/src/lib.rs:7]: see todo

File: crates/yulgen/src/mappers/expressions.rs [hunk 4]
```diff
@@ -25,6 +25,7 @@ use yultsur::*;
 pub fn expr(context: &mut FnContext, exp: &Node<fe::Expr>) -> yul::Expression {
     let expression = match &exp.kind {
         fe::Expr::Name(_) => expr_name(exp),
+        fe::Expr::Path(_) => todo!(), // XXX
```
Comment (g-r-a-n-t) [inline crates/yulgen/src/mappers/expressions.rs:28]: dont think there's a case where we generate Yul from a path. constants are lowered away

File: crates/analyzer/src/db/queries/module.rs [hunk 5]
```diff
@@ -190,3 +199,166 @@ pub fn module_constant_type(
...
+    tree: ast::UseTree,
+) -> Analysis<Rc<IndexMap<String, Item>>> {
+    match tree {
+        ast::UseTree::Glob { prefix } => {
+            let prefix_module = Item::Module(module).resolve_path(db, &prefix);
+
+            let items = match prefix_module.value {
+                Some(Item::Module(module)) => module.items(db),
+                Some(_) => {
+                    panic!("not a module");
```
Comment (g-r-a-n-t) [inline crates/analyzer/src/db/queries/module.rs:248]: should be a diagnostic

File: crates/analyzer/src/db/queries/module.rs [hunk 6]
```diff
@@ -190,3 +199,166 @@ pub fn module_constant_type(
...
+                    let child_items = db.module_resolve_use_tree(module, node.kind.clone());
+                    diagnostics.append(&mut (*child_items.diagnostics).clone());
+
+                    accum
+                        .into_iter()
+                        .chain((*child_items.value).clone())
+                        .collect::<IndexMap<_, _>>()
+                }),
+                Some(_) => {
+                    panic!("not a module");
```
Comment (g-r-a-n-t) [inline crates/analyzer/src/db/queries/module.rs:274]: diagnostic

File: crates/analyzer/src/namespace/items.rs [hunk 7]
```diff
@@ -96,16 +149,95 @@ impl Item {
...
+
+    pub fn name(&self, db: &dyn AnalyzerDb) -> String {
+        self.data(db).name.clone()
+    }
+
+    pub fn main_module(&self, db: &dyn AnalyzerDb) -> ModuleId {
+        *self
+            .all_modules(db)
+            .iter()
+            // TODO: finding by name is not valid
```
Comment (g-r-a-n-t) [inline crates/analyzer/src/namespace/items.rs:193]: fix this

File: crates/analyzer/src/traversal/expressions.rs [hunk 8]
```diff
@@ -832,80 +861,116 @@ fn expr_call_name(
...
+        ))
+    })?;
+
+    expr_call_named_thing(scope, named_thing, node, generic_args, args)
+}
+
+fn expr_call_named_thing<T: std::fmt::Display>(
+    scope: &mut BlockScope,
+    named_thing: NamedThing,
+    node: &Node<T>,
```
Comment (g-r-a-n-t) [inline crates/analyzer/src/traversal/expressions.rs:894]: rename to `func_node`

File: crates/driver/src/lib.rs [hunk 9]
```diff
@@ -62,18 +78,128 @@ pub fn compile(
...
+        return Err(CompileError(errors));
+    }
+
+    let lowered_ingot_id = fe_lowering::lower_ingot(&db, ingot_id);
+
+    fe_analyzer::analyze_ingot(&db, lowered_ingot_id).expect("failed to analyze lowered AST");
+
+    let lowered_module_id = lowered_ingot_id.main_module(&db);
+
+    // TODO: fix this
```
Comment (g-r-a-n-t) [inline crates/driver/src/lib.rs:197]: fix this

File: crates/driver/src/lib.rs [hunk 10]
```diff
@@ -122,11 +247,66 @@ pub fn compile(
...
+        .collect::<IndexMap<_, _>>();
 
     Ok(CompiledModule {
         src_ast,
         lowered_ast,
         contracts,
     })
 }
+
+// fn validate_ingot_file_paths(
```
Comment (g-r-a-n-t) [inline crates/driver/src/lib.rs:259]: finish implementing ingot validation

Comment (g-r-a-n-t) [reply crates/driver/src/lib.rs:259]: should prly make the ingot validation part of salsa

File: crates/analyzer/src/namespace/items.rs [hunk 11]
```diff
@@ -116,6 +246,39 @@ impl ModuleId {
...
+        self.data(db).file_content.clone()
+    }
+
+    pub fn ingot_path(&self, db: &dyn AnalyzerDb) -> String {
+        match self.context(db) {
+            ModuleContext::Ingot(ingot) => match self.file_content(db) {
+                ModuleFileContent::Dir { dir_path } => dir_path,
+                ModuleFileContent::File { file } => ingot.data(db).fe_files[&file].0.name.clone(),
+            },
+            ModuleContext::Global(_) => panic!("cant do this"),
```
Comment (g-r-a-n-t) [inline crates/analyzer/src/namespace/items.rs:263]: better panic message

File: crates/analyzer/tests/analysis.rs [hunk 12]
```diff
@@ -46,10 +60,60 @@ macro_rules! test_analysis {
...
+        fn $name() {
+            let files = test_files::build_filestore($path);
+
+            let db = TestDb::default();
+
+            let global = Global::default();
+            let global_id = db.intern_global(Rc::new(global));
+
+            let ingot = Ingot {
+                name: "basic_ingot".to_string(),
```
Comment (g-r-a-n-t) [inline crates/analyzer/tests/analysis.rs:85]: fix ingot name

File: crates/analyzer/tests/analysis.rs [hunk 13]
```diff
@@ -224,6 +290,8 @@ fn build_snapshot(file_store: &FileStore, module: items::ModuleId, db: &dyn Anal
             | Item::GenericType(_)
             | Item::BuiltinFunction(_)
             | Item::Object(_) => vec![],
+            Item::Ingot(_) => todo!(),
```
Comment (g-r-a-n-t) [inline crates/analyzer/tests/analysis.rs:293]: fix

File: crates/analyzer/tests/snapshots/analysis__basic_ingot.snap [hunk 14]
```diff
@@ -0,0 +1,195 @@
...
+3 │   my_u256: u256
+  │   ^^^^^^^^^^^^^ u256
+4 │   my_i8: i8
+  │   ^^^^^^^^^ i8
+
+
+note: 
+  ┌─ ingots/basic_ingot/src/ding/dang.fe:1:1
+  │
+1 │ type Dang = u256[42]
```
Comment (g-r-a-n-t) [inline crates/analyzer/tests/snapshots/analysis__basic_ingot.snap:148]: look into this

File: crates/driver/src/lib.rs [hunk 15]
```diff
@@ -62,18 +76,130 @@ pub fn compile(
...
+        lowered_ast,
+        contracts,
+    })
+}
+
+/// Compiles the given Fe source code to all targets.
+///
+/// If `with_bytecode` is set to false, the compiler will skip the final Yul ->
+/// Bytecode pass. This is useful when debugging invalid Yul code.
+pub fn compile_ingot(
```
Comment (g-r-a-n-t) [inline crates/driver/src/lib.rs:145]: update comment

File: crates/driver/src/lib.rs [hunk 16]
```diff
@@ -62,18 +76,130 @@ pub fn compile(
...
+    files: &FileStore,
+    file_ids: &[SourceFileId],
+    _with_bytecode: bool,
+    _optimize: bool,
+) -> Result<CompiledModule, CompileError> {
+    // parse source
```
Comment (g-r-a-n-t) [inline crates/driver/src/lib.rs:151]: remove comment

File: crates/driver/src/lib.rs [hunk 17]
```diff
@@ -62,18 +76,130 @@ pub fn compile(
...
+    let global_id = db.intern_global(Rc::new(global));
+
+    let ingot = Ingot {
+        name: "".to_string(),
+        global: global_id,
+        fe_files: file_ids
+            .iter()
+            .map(|file_id| {
+                let file = files.get_file(*file_id).unwrap();
+                let ast = parse_file(*file_id, &file.content).unwrap().0;
```
Comment (g-r-a-n-t) [inline crates/driver/src/lib.rs:167]: looks like we're throwing away diagnostics

File: crates/fe/src/main.rs [hunk 18]
```diff
@@ -112,6 +134,23 @@ pub fn main() {
     }
 }
 
+fn build_filestore_for_dir(path: &str) -> FileStore {
```
Comment (g-r-a-n-t) [inline crates/fe/src/main.rs:137]: should handle missing path

File: crates/lowering/src/lib.rs [hunk 19]
```diff
@@ -1,15 +1,20 @@
...
 mod mappers;
 mod names;
 mod utils;
 
+pub use db::{LoweringDb, TestDb};
+
 /// Lowers the Fe source AST to a Fe HIR AST.
-pub fn lower(db: &dyn AnalyzerDb, module: ModuleId) -> ast::Module {
-    mappers::module::module(db, module)
+pub fn lower_module(db: &dyn LoweringDb, module_id: ModuleId) -> ModuleId {
```
Comment (g-r-a-n-t) [inline crates/lowering/src/lib.rs:14]: fix above comment

File: crates/parser/src/grammar/module.rs [hunk 20]
```diff
@@ -130,60 +130,68 @@ pub fn parse_use(par: &mut Parser) -> ParseResult<Node<Use>> {
     Ok(Node::new(Use { tree }, use_tok.span + tree_span))
 }
 
-/// Parse a `::` delimited path.
-pub fn parse_path(par: &mut Parser) -> ParseResult<Node<Path>> {
-    let mut names = vec![];
+// /// Parse a `::` delimited path.
```
Comment (g-r-a-n-t) [inline crates/parser/src/grammar/module.rs:133]: delete this code

File: crates/analyzer/src/namespace/items.rs [hunk 21]
```diff
@@ -155,19 +339,16 @@ impl ModuleId {
...
+                    sink.push(&diag)
                 }
-                ast::ModuleStmt::Use(inner) => {
-                    sink.push(&errors::not_yet_implemented("use", inner.span));
-                }
-                _ => {} // everything else is a type def, handled below.
             }
         }
 
+        // TODO: should get diagnostics from item map query
```
Comment (g-r-a-n-t) [inline crates/analyzer/src/namespace/items.rs:349]: note comment

File: crates/analyzer/src/db/queries/module.rs [hunk 22]
```diff
@@ -190,3 +226,228 @@ pub fn module_constant_type(
...
+) -> Analysis<Rc<IndexMap<String, (Span, Item)>>> {
+    let mut diagnostics = vec![];
+
+    let ast::Module { body } = &module.data(db).ast;
+
+    let items = body
+        .iter()
+        .fold(indexmap! {}, |mut accum, stmt| {
+            if let ast::ModuleStmt::Use(use_stmt) = stmt {
+                let parent = module.parent_module(db).unwrap();
```
Comment (g-r-a-n-t) [inline crates/analyzer/src/db/queries/module.rs:242]: expect

File: crates/analyzer/src/db/queries/module.rs [hunk 23]
```diff
@@ -190,3 +226,228 @@ pub fn module_constant_type(
...
+                let items = db.module_resolve_use_tree(parent, use_stmt.kind.tree.clone());
+                diagnostics.append(&mut (*items.diagnostics).clone());
```
Comment (g-r-a-n-t) [inline crates/analyzer/src/db/queries/module.rs:244]: extend

File: crates/analyzer/src/db/queries/module.rs [hunk 24]
```diff
@@ -190,3 +226,228 @@ pub fn module_constant_type(
...
+
+                for (name, (tree_span, item)) in items.value.iter() {
+                    if let Some((other_tree_span, other_item)) =
```
Comment (g-r-a-n-t) [inline crates/analyzer/src/db/queries/module.rs:247]: should use the final segment or rename span

File: crates/analyzer/src/db/queries/module.rs [hunk 25]
```diff
@@ -190,3 +226,228 @@ pub fn module_constant_type(
...
+        value: Rc::new(items),
+        diagnostics: Rc::new(diagnostics),
+    }
+}
+
+pub fn module_resolve_use_tree(
+    db: &dyn AnalyzerDb,
+    module: ModuleId,
+    tree: Node<ast::UseTree>,
+) -> Analysis<Rc<IndexMap<String, (Span, Item)>>> {
```
Comment (g-r-a-n-t) [inline crates/analyzer/src/db/queries/module.rs:295]: comments on all these methods describing the return value

File: crates/analyzer/src/namespace/items.rs [hunk 26]
```diff
@@ -116,6 +246,39 @@ impl ModuleId {
...
+        match self.context(db) {
+            ModuleContext::Ingot(ingot) => match self.file_content(db) {
+                ModuleFileContent::Dir { dir_path } => dir_path,
+                ModuleFileContent::File { file } => ingot.data(db).fe_files[&file].0.name.clone(),
+            },
+            ModuleContext::Global(_) => panic!("cannot get path"),
+        }
+    }
+
+    pub fn has_dir(&self, db: &dyn AnalyzerDb) -> bool {
```
Comment (g-r-a-n-t) [inline crates/analyzer/src/namespace/items.rs:267]: delete this

File: crates/analyzer/src/namespace/items.rs [hunk 27]
```diff
@@ -126,21 +289,42 @@ impl ModuleId {
         db.module_all_items(*self)
     }
 
-    pub fn imported_items(&self, db: &dyn AnalyzerDb) -> Rc<IndexMap<String, Item>> {
-        db.module_imported_item_map(*self)
+    pub fn used_items(&self, db: &dyn AnalyzerDb) -> Rc<IndexMap<String, (Span, Item)>> {
```
Comment (g-r-a-n-t) [inline crates/analyzer/src/namespace/items.rs:292]: comment on this

File: crates/analyzer/src/namespace/items.rs [hunk 28]
```diff
@@ -126,21 +289,42 @@ impl ModuleId {
...
+        db.module_used_item_map(*self).value
     }
 
     pub fn resolve_name(&self, db: &dyn AnalyzerDb, name: &str) -> Option<Item> {
         self.items(db)
             .get(name)
             .copied()
-            .or_else(|| self.imported_items(db).get(name).copied())
+            .or_else(|| self.used_items(db).get(name).map(|(_, item)| *item))
```
Comment (g-r-a-n-t) [inline crates/analyzer/src/namespace/items.rs:300]: is this necessary?

File: crates/driver/src/lib.rs [hunk 29]
```diff
@@ -24,29 +28,44 @@ pub struct CompiledContract {
 #[derive(Debug)]
 pub struct CompileError(pub Vec<Diagnostic>);
 
-/// Compiles the given Fe source code to all targets.
+/// Compiles a single input file.
 ///
 /// If `with_bytecode` is set to false, the compiler will skip the final Yul ->
 /// Bytecode pass. This is useful when debugging invalid Yul code.
-pub fn compile(
+pub fn compile_module(
```
Comment (g-r-a-n-t) [inline crates/driver/src/lib.rs:35]: should pass in names to both driver funcs

**Comment by sbillig** [comment]
:eyes:

**Comment by g-r-a-n-t** [comment]
> eyes

yup, feel free to look

**Review by sbillig — APPROVED** [review]
Sweet, looks good to me (I sorta skimmed over the filesystem stuff).

Merge baby merge.

**Pull request merged**